server/controllers/bookController.js

```diff
@@ -71,68 +58,92 @@ class BookController {
-        return Utils.errorStatus(res, 400, responseObj); 
+    const booksNotReturned = await models.BorrowedBooks.findAndCountAll({ where: { patronId, returned : false } });
+
+    if (booksNotReturned.count >= 3) {
+      return Utils.errorStatus(res, 400, 'You cannot have more than 3 books in your possession');        
+    } 
+    const today = new Date();
+    const overdueBook = booksNotReturned.rows.filter((book) => book.dueDate < today);
+
+    if(overdueBook.length >= 1) {
+      const dueBooks = overdueBook.map((book) =>`The book titled '${book.title.toUpperCase()}' was due for return on ${book.dueDate.toString().substring(0, 15)},`);
+      const dueBookTwo = dueBooks[1] ? dueBooks[1] : '' ;
+      const responseObj = {
+        message: `${dueBooks[0]} ${dueBookTwo} Kindly return before we can proceed with a new request`
       }
+      return Utils.errorStatus(res, 400, responseObj); 
+    }
 
-      const dateBorrowed = new Date();
-      const dueDate = new Date();
-      // Books are due for return after 3 days
-      dueDate.setDate(dueDate.getDate() + 3); 
-      const bookDetails = await models.Books.findOne({ where: { isbn }});
-      const { title } = bookDetails.dataValues ;
-      const borrowedBook = await models.BorrowedBooks.create({
-        isbn,
-        title,
-        dateBorrowed,
-        dueDate,
-        patronId    
-      });
+    const dateBorrowed = new Date();
+    const dueDate = new Date();
+    // Books are due for return after 3 days
+    dueDate.setDate(dueDate.getDate() + 3); 
+    const bookDetails = await models.Books.findOne({ where: { isbn }});
+    const { title } = bookDetails.dataValues ;
+    const borrowedBook = await models.BorrowedBooks.create({
+      isbn,
+      title,
+      dateBorrowed,
+      dueDate,
+      patronId    
+    });
+
+    await models.Books.update({ status: 'borrowed' }, { where: { isbn } });
+    return Utils.successStatus(res, 201, 'Success', borrowedBook);
+  }
 
-      await models.Books.update({ status: 'borrowed' }, { where: { isbn } });
+  static async extendDueDate(req, res) {
+    const { id } = req.params;
+    const { date } = req.body;
+    const book = await models.BorrowedBooks.findByPk(id);
 
-      return Utils.successStatus(res, 201, 'Success', borrowedBook);
+    if(req.loggedinUser.id !== book.patronId) {
+      return Utils.errorStatus(res, 400, `You do not have any book with id: ${id} in your possession`);
+    }
+
+    if(new Date() > book.dueDate) {
+      return Utils.errorStatus(
+        res,
+        400,
+        'You cannot extend the borrowing period when the book is overdue. Kindly return the book to the library'
+      );
+    }
 
-    } catch (error) {
-      Utils.errorStatus(res, 500, error.message);
+    if(new Date(date) < book.dueDate) {
+      return Utils.errorStatus(
+        res,
+        400,
+        'The new date must be a date later than the previous due date'
+      );
     }
+
+    const newDate = new Date(date);
```

Comment:
Your story [#166265868](https://www.pivotaltracker.com/story/show/166265868) stated that _There should be a limit to the number of times one can extend._ I think you missed that validation.

Reply:
Check this out.
[#166265899](https://www.pivotaltracker.com/story/show/166265899)